src/core/__tests__/ReactComponent-test.js

```diff
@@ -165,47 +142,108 @@ describe('ReactComponent', function() {
             title="wrapper"
-            ref={this.wrapperRef}
+            ref={(c) => this.wrapperRef = c}
             getContent={this.getInner}
             />
         );
       },
       componentDidMount: function() {
-        // TODO: Currently new refs aren't available on initial render
-      },
-      componentDidUpdate: function() {
         // Check .props.title to make sure we got the right elements back
-        this.wrapperRef.then(function(wrapper) {
-          expect(wrapper.props.title).toBe("wrapper");
-          refsResolved++;
-        });
-        this.innerRef.then(function(inner) {
-          expect(inner.props.title).toEqual("inner");
-          refsResolved++;
-        });
-        expect(refsResolved).toBe(0);
+        expect(this.wrapperRef.props.title).toBe('wrapper');
+        expect(this.innerRef.props.title).toBe('inner');
+        mounted = true;
       }
     });
 
     var instance = <Component />;
     instance = ReactTestUtils.renderIntoDocument(instance);
-    instance.forceUpdate();
-    expect(refsResolved).toBe(2);
+    expect(mounted).toBe(true);
+  });
+
+  it('should call refs at the correct time', function() {
+    var log = [];
+
+    var Inner = React.createClass({
+      render: function() {
+        return <div />;
+      },
+      componentDidMount: function() {
+        log.push(`inner ${this.props.id} componentDidMount`);
+      },
+      componentDidUpdate: function() {
+        log.push(`inner ${this.props.id} componentDidUpdate`);
+      },
+      componentWillUnmount: function() {
+        log.push(`inner ${this.props.id} componentWillUnmount`);
+      }
+    });
+
+    var Outer = React.createClass({
+      render: function() {
+        return (
+          <div>
+            <Inner id={1} ref={(c) => {
+              log.push(`ref 1 got ${c ? `instance ${c.props.id}` : 'null'}`);
+            }}/>
+            <Inner id={2} ref={(c) => {
+              log.push(`ref 2 got ${c ? `instance ${c.props.id}` : 'null'}`);
+            }}/>
+          </div>
+        );
+      },
+      componentDidMount: function() {
+        log.push('outer componentDidMount');
+      },
+      componentDidUpdate: function() {
+        log.push('outer componentDidUpdate');
+      },
+      componentWillUnmount: function() {
+        log.push('outer componentWillUnmount');
+      }
+    });
+
+    // mount, update, unmount
+    var el = document.createElement('div');
+    log.push('start mount');
+    React.render(<Outer />, el);
+    log.push('start update');
+    React.render(<Outer />, el);
+    log.push('start unmount');
+    React.unmountComponentAtNode(el);
+
+    expect(log).toEqual([
+      'start mount',
+      'inner 1 componentDidMount',
+      'ref 1 got instance 1',
+      'inner 2 componentDidMount',
+      'ref 2 got instance 2',
+      'outer componentDidMount',
+      'start update',
+      'inner 1 componentDidUpdate',
+      'ref 1 got null',  // Previous (equivalent) ref gets cleared first
```

Comment:
This is kind of weird. If we do keep this, maybe it makes more sense to null out the ref before componentWillReceiveProps/shouldComponentUpdate/componentWillUpdate.